### PR TITLE
Add reactivity calculation for common fusion reactions 

### PR DESCRIPTION
**Pull request by CoronelBuendia**
## Linked Issues

Closes #{ID}

## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `flake8` and `black .`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/plasma_physics/reactions.py [hunk 1]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
     return n_DT_reactions(p_fus) * D_MOLAR_MASS / N_AVOGADRO
+
+
+REACTIONS = {
+    "D-T": 0,  # D + T --> 4He + n reaction
+    "D-D": -1,  # D + D --> 0.5 D-D1 + 0.5 D-D2
+    "D-D1": 1,  # D + D --> 3He + n reaction [50 %]
+    "D-D2": 2,  # D + D --> T + p reaction [50 %]
+    "D-He3": 3,  # D + 3He --> 4He + p reaction
+}
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:189]: This feels very enum-y

```suggestion
class Reactions(Enum):
    D_T =  auto(),  # D + T --> 4He + n reaction
    D_D = auto(),  # D + D --> 0.5 D-D1 + 0.5 D-D2
    D_D1 = auto(),  # D + D --> 3He + n reaction [50 %]
    D_D2 =  auto(),  # D + D --> T + p reaction [50 %]
    D_He3 = auto(),  # D + 3He --> 4He + p reaction

```

File: bluemira/plasma_physics/reactions.py [hunk 2]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+    method: str
+        The parameterisation to use when calculating the reactivity
+
+    Returns
+    -------
+    sigma_v: float
+        Reactivity of the reaction at the specified temperature(s) [m^3/s]
+    """
+    if reaction not in REACTIONS:
+        raise ValueError(f"Unknown reaction: {reaction}")
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:214]:
```suggestion
    reaction = Reactions[reaction.replace('-','_')]
```

File: bluemira/plasma_physics/reactions.py [hunk 3]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+    -------
+    sigma_v: float
+        Reactivity of the reaction at the specified temperature(s) [m^3/s]
+
+    Notes
+    -----
+    H.-S. Bosch and G.M. Hale 1992 Nucl. Fusion 32 611
+    DOI 10.1088/0029-5515/32/4/I07
+    """
+    if reaction == "D-D":
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:361]:
```suggestion
    if reaction == Reactions.D_D:
```

File: bluemira/plasma_physics/reactions.py [hunk 4]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+        return 0.5 * (
+            _reactivity_bosch_hale(temp_kev, "D-D1")
+            + _reactivity_bosch_hale(temp_kev, "D-D2")
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:364]:
```suggestion
            _reactivity_bosch_hale(temp_kev, Reactions.D_D1)
            + _reactivity_bosch_hale(temp_kev, Reactions.D_D2)
```

File: bluemira/plasma_physics/reactions.py [hunk 5]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+        )
+    mapping = {
+        "D-T": BoschHale_DT_4Hen,
+        "D-D1": BoschHale_DD_3Hen,
+        "D-D2": BoschHale_DD_Tp,
+        "D-He3": BoschHale_DHe3_4Hep,
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:370]:
```suggestion
        Reactions.D_T: BoschHale_DT_4Hen,
        Reactions.D_D1: BoschHale_DD_3Hen,
        Reactions.D_D2: BoschHale_DD_Tp,
        Reactions.D_He3: BoschHale_DHe3_4Hep,
```

File: bluemira/plasma_physics/reactions.py [hunk 6]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+        Temperature [keV]
+    reaction: str
+        The fusion reaction
+
+    Returns
+    -------
+    sigma_v: float
+        Reactivity of the reaction at the specified temperature(s) [m^3/s]
+    """
+    if reaction == "D-T":
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:417]:
```suggestion
    if reaction == Reactions.D_T:
```

File: bluemira/plasma_physics/reactions.py [hunk 7]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+        t3 = temp_kev ** (-1 / 3)
+
+        term_1 = 8.972 * np.exp(-19.9826 * t3) * t3**2
+        term_2 = (temp_kev + 1.0134) / (1 + 6.386e-3 * (temp_kev + 1.0134) ** 2)
+        term_3 = 1.877 * np.exp(-0.16176 * temp_kev * np.sqrt(temp_kev))
+        return 1e-19 * term_1 * (term_2 + term_3)
+
+    elif reaction == "D-D":
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:425]:
```suggestion
    elif reaction == Reactions.D_D:
```

File: bluemira/plasma_physics/reactions.py [hunk 8]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+    -------
+    sigma_v: float
+        Reactivity of the reaction at the specified temperature(s) [m^3/s]
+
+    Notes
+    -----
+    Johner, Jean (2011). HELIOS: a zero-dimensional tool for next step and reactor
+    studies. Fusion Science and Technology, 59(2), 308-313. Appendix E.II
+    """
+    if reaction != "D-T":
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:456]:
```suggestion
    if reaction != Reactions.D_T:
```

File: bluemira/plasma_physics/reactions.py [hunk 9]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+        raise ValueError(f"This function only supports D-T, not {reaction}")
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:457]:
```suggestion
        raise ValueError(f"This function only supports D-T, not {reaction.name.replace('_','-')}")
```

File: bluemira/plasma_physics/reactions.py [hunk 10]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+        return 1e-19 * term_1 * (term_2 + term_3)
+
+    elif reaction == "D-D":
+        term_1 = (
+            0.16247 + 0.001741 * temp_kev - 0.029 * np.exp(-0.3843 * np.sqrt(temp_kev))
+        )
+        term_2 = np.exp(-18.8085 / (temp_kev ** (1 / 3))) / (temp_kev ** (1 / 3)) ** 2
+        return 1e-19 * term_1 * term_2
+    else:
+        raise ValueError(f"This function only supports D-D and D-T, not {reaction}")
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:432]:
```suggestion
        raise ValueError(f"This function only supports D-D and D-T, not {reaction.name.replace('_','-')}")
```

File: bluemira/plasma_physics/reactions.py [hunk 11]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+    return (
+        1e-6
+        * data.c[0]
+        * theta
+        * np.sqrt(chi / (data.mrc2 * temp_kev**3))
+        * np.exp(-3 * chi)
+    )
+
+
+def _reactivity_plasmod(temp_kev, reaction):
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:401]:
```suggestion
def _reactivity_plasmod(temp_kev: Union[float, np.ndarray], reaction: Reactions) -> Union[float, np.ndarray]:
```

File: bluemira/plasma_physics/reactions.py [hunk 12]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+    """
+    Reactivity equations used in PLASMOD (original source unknown)
+
+    Parameters
+    ----------
+    temp_kev: Union[float, np.ndarray]
+        Temperature [keV]
+    reaction: str
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:409]:
```suggestion
    reaction: Reactions
```

File: bluemira/plasma_physics/reactions.py [hunk 13]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+            -2.02896e-3,
+            -1.91080e-5,
+            1.35776e-4,
+            0.0,
+            0.0,
+        ]
+    )
+
+
+def _reactivity_bosch_hale(temp_kev, reaction):
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:340]:
```suggestion
def _reactivity_bosch_hale(temp_kev: Union[float, np.ndarray], reaction: Reactions) -> Union[float, np.ndarray]:
```

File: bluemira/plasma_physics/reactions.py [hunk 14]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+        term_1 = (
+            0.16247 + 0.001741 * temp_kev - 0.029 * np.exp(-0.3843 * np.sqrt(temp_kev))
+        )
+        term_2 = np.exp(-18.8085 / (temp_kev ** (1 / 3))) / (temp_kev ** (1 / 3)) ** 2
+        return 1e-19 * term_1 * term_2
+    else:
+        raise ValueError(f"This function only supports D-D and D-T, not {reaction}")
+
+
+def _reactivity_johner(temp_kev, reaction):
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:435]:
```suggestion
def _reactivity_johner(temp_kev: Union[float, np.ndarray], reaction: Reactions) -> Union[float, np.ndarray]:
```

File: bluemira/plasma_physics/reactions.py [hunk 15]
```diff
@@ -174,3 +178,302 @@ def r_D_burn_DT(p_fus):  # noqa :N802
...
+    """
+    Johner's monomial fit for analytical calculations
+
+    Parameters
+    ----------
+    temp_kev: Union[float, np.ndarray]
+        Temperature [keV]
+    reaction: str
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:443]:
```suggestion
    reaction: Reactions
```

File: bluemira/plasma_physics/reactions.py [hunk 16]
```diff
@@ -23,6 +23,9 @@
 Fusion reactions
 """
 
+from dataclasses import dataclass
+from typing import Union
```
Comment (je-cook) [inline bluemira/plasma_physics/reactions.py:27]:
```suggestion
from dataclasses import dataclass
from enum import Enum, auto
from typing import Union
```

**Review by je-cook — CHANGES_REQUESTED** [review]
I felt that `REACTIONS` should be an enum. I've done that and added a bit of typing.

My only other though is that this not in our units and is user facing but is in keV. Could we have a line at the top of the reactivity function that converts it to keV as I assume all the relations need that?

```python
def reactivity
.
.
    temp_kev = raw_uc(temp_ev, 'eV', 'keV')
```

**Comment by CoronelBuendia** [comment]
> I felt that `REACTIONS` should be an enum. I've done that and added a bit of typing.
> 
> My only other though is that this not in our units and is user facing but is in keV. Could we have a line at the top of the reactivity function that converts it to keV as I assume all the relations need that?
> 
> ```python
> def reactivity
> .
> .
>     temp_kev = raw_uc(temp_ev, 'eV', 'keV')
> ```

Shouldn't it technically be in [K] then? [keV] is really ubiquitous... eV not so much!

**Comment by je-cook** [comment]
> > I felt that `REACTIONS` should be an enum. I've done that and added a bit of typing.
> > My only other though is that this not in our units and is user facing but is in keV. Could we have a line at the top of the reactivity function that converts it to keV as I assume all the relations need that?
> > ```python
> > def reactivity
> > .
> > .
> >     temp_kev = raw_uc(temp_ev, 'eV', 'keV')
> > ```
> 
> Shouldn't it technically be in [K] then? [keV] is really ubiquitous... eV not so much!

Yep you're right

**Review by je-cook — APPROVED** [review]

**Pull request merged**